File: apps/api/app/schemas/obligation.py
from pydantic import BaseModel, Field


class CreateObligationInput(BaseModel):
    name: str = Field(min_length=1)
    categoryId: str
    estimatedAmount: float = Field(gt=0)
    currency: str = Field(min_length=3, max_length=3)
    dueDay: int = Field(ge=1, le=31)


class UpdateObligationInput(BaseModel):
    name: str | None = Field(default=None, min_length=1)
    categoryId: str | None = None
    estimatedAmount: float | None = Field(default=None, gt=0)
    dueDay: int | None = Field(default=None, ge=1, le=31)


class LinkMovementInput(BaseModel):
    movementId: str | None = None


class ObligationResponse(BaseModel):
    id: str
    name: str
    categoryId: str
    estimatedAmount: float
    baseAmount: float | None = None
    carryoverAmount: float | None = None
    currency: str
    dueDay: int
    isPaid: bool
    linkedMovementId: str | None = None
    linkedMovementAmount: float | None = None
    isActive: bool
    createdAt: str
    updatedAt: str


class ObligationSummaryResponse(BaseModel):
    currency: str
    totalObligations: float
    paidAmount: float
    pendingAmount: float
    coveragePercent: float
    currentBalance: float
    freeAfterObligations: float


class ObligationHistoryItemResponse(BaseModel):
    obligationId: str
    name: str
    currency: str
    dueAmount: float
    paidAmount: float
    isPaid: bool


class ObligationHistoryMonthResponse(BaseModel):
    month: str
    monthLabel: str
    totalDue: float
    totalPaid: float
    items: list[ObligationHistoryItemResponse]

File: packages/types/src/obligation.ts
import type { Currency } from "./account";

export interface Obligation {
  id: string;
  name: string;
  bucket: "necessity" | "desire" | "save_invest";
  categoryId: string;
  estimatedAmount: number;
  baseAmount?: number;
  carryoverAmount?: number;
  currency: Currency;
  dueDay: number;
  isPaid: boolean;
  linkedMovementId?: string;
  linkedMovementAmount?: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CreateObligationInput {
  name: string;
  bucket: "necessity" | "desire" | "save_invest";
  categoryId: string;
  estimatedAmount: number;
  currency: Currency;
  dueDay: number;
}

export interface UpdateObligationInput {
  name?: string;
  bucket?: "necessity" | "desire" | "save_invest";
  categoryId?: string;
  estimatedAmount?: number;
  dueDay?: number;
}

export interface ObligationSummary {
  currency: Currency;
  totalObligations: number;
  paidAmount: number;
  pendingAmount: number;
  coveragePercent: number;
  currentBalance: number;
  freeAfterObligations: number;
}

export interface ObligationHistoryItem {
  obligationId: string;
  name: string;
  currency: Currency;
  dueAmount: number;
  paidAmount: number;
  isPaid: boolean;
}

export interface ObligationHistoryMonth {
  month: string;
  monthLabel: string;
  totalDue: number;
  totalPaid: number;
  items: ObligationHistoryItem[];
}
